cpp/src/slice2cs/CsUtil.cpp

```diff
@@ -1864,212 +1865,91 @@ Slice::CsGenerator::toArrayAlloc(const string& decl, const string& sz)
-                StructPtr st = dynamic_pointer_cast<Struct>(dataMember->container());
-                ExceptionPtr ex = dynamic_pointer_cast<Exception>(dataMember->container());
-                ClassDeclPtr cl = dynamic_pointer_cast<ClassDecl>(dataMember->container());
-                static const string csTypePrefix = "cs:type:";
-                // TODO Seems like this validation only got half written?...
-            }
-            else if (dynamic_pointer_cast<Module>(cont))
+        }
+        else if (dynamic_pointer_cast<Dictionary>(p))
+        {
+            if (argument != "SortedDictionary" && argument != "SortedList")
             {
-                if (directive == "cs:namespace" && !arguments.empty())
-                {
-                    newLocalMetadata.push_back(metadata);
-                    continue;
-                }
+                return "the 'cs:generic' metadata only supports 'SortedDictionary' and 'SortedList' as arguments when "
+                       "applied to a dictionary";
             }
+        }
+        return nullopt;
+    };
+    knownMetadata.emplace("cs:generic", genericInfo);
 
-            if (directive == "cs:attribute" && !arguments.empty())
+    // TODO: this is being removed. We tell the validator to perform basically no checking.
+    // "cs:implements"
+    MetadataInfo implementsInfo;
+    implementsInfo.acceptedArgumentKind = MetadataArgumentKind::OptionalTextArgument;
+    implementsInfo.mustBeUnique = false;
+    knownMetadata.emplace("cs:implements", std::move(implementsInfo));
+
+    // "cs:namespace"
+    MetadataInfo namespaceInfo;
+    namespaceInfo.validOn = {typeid(Module)};
+    namespaceInfo.acceptedArgumentKind = MetadataArgumentKind::SingleArgument;
+    namespaceInfo.extraValidation = [](const MetadataPtr&, const SyntaxTreeBasePtr& p) -> optional<string>
+    {
+        // 'cs:namespace' can only be applied to top-level modules
+        if (auto mod = dynamic_pointer_cast<Module>(p))
+        {
+            // Top-level modules are contained by the 'Unit'. Non-top-level modules are contained in 'Module's.
+            if (!dynamic_pointer_cast<Unit>(mod->container()))
             {
-                newLocalMetadata.push_back(metadata);
-                continue;
+                return "the 'cs:namespace' metadata can only be applied to top-level modules";
             }
-
-            ostringstream msg;
-            msg << "ignoring invalid metadata '" << *metadata << "'";
-            cont->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, msg.str());
-            continue;
         }
-        newLocalMetadata.push_back(metadata);
-    }
+        return nullopt;
+    };
+    knownMetadata.emplace("cs:namespace", std::move(namespaceInfo));
+
+    // "cs:property"
+    MetadataInfo propertyInfo;
+    propertyInfo.validOn = {typeid(ClassDecl), typeid(Slice::Exception), typeid(Struct)};
```

Comment:
For validOn, we must used ClassDecl/InterfaceDecl and not ClassDef/InterfaceDef?

Reply:
Yes, after my PR for simplifying the Decl/Def situation, you must always pass in the Decls.
There's a comment saying so in `MetadataValidator.h`:
https://github.com/zeroc-ice/ice/blob/6e237b53f7d94e64ff1b5cddde6504319226dac2/cpp/src/Slice/MetadataValidation.h#L68-L70